### ISSUE TITLE
Desired behavior when casting `Tensor` to cuda?

### ISSUE THREAD

**Issue by juesato** [torch/cutorch#511]
Right now, if I cast a `ByteTensor` to cuda, as in
```
a = torch.ByteTensor()
a:cuda()
```
it will have type `CudaTensor` rather than `CudaByteTensor`. Because some modules (like `maskedFill`) require a `CudaByteTensor`, it's rather awkward getting an object of type `CudaByteTensor`. 

- Is this the desired behavior? (I'm not sure if the reason the conversion is done this way is because not all operations have been implemented for different Tensor types, or maybe another reason)
- Is there a way to easily get around this i.e. construct a Tensor which will be a `ByteTensor` on CPU, and a `CudaByteTensor` on GPU?

**Comment by soumith** [comment]
this is unfortunate, and cant be changed because of backward compatibility.
This probably needs a new function that retains the type, like maybe :cuda2() or something, but i dont know of a decent name / path for this..

**Comment by fmassa** [comment]
Yes, that's unfortunate, and dates back from the time cutorch didn't have other tensor typed than float.
When creating a new module (like CMaxTable), it seems then that you might need to overwrite the `type` function for the module, and add a special treatment for cuda tensors, as done for example in `nn.Max`  or in here https://github.com/torch/nn/blob/master/Normalize.lua#L26

**Comment by juesato** [comment]
Thanks for both those tips - that's very helpful. I think I'll go with constructing a `CudaByteTensor` similar to how the `Normalize` layer implements it. It seems like this doesn't completely fix the problem, since then `gt` expects the two inputs to both be `CudaByteTensor`s if the second argument isn't a constant, but I think I can get around this by using two masks and copying between the two (suboptimal, but at least doesn't allocate memory)

**Comment by tom-roddick** [comment]
I came across this issue recently, and was wondering whether now that there several different types of Cuda tensors, it would be worth having `TYPETensor:gpu()` and corresponding `CudaTYPETensor:cpu()` functions which convert to the appropriate Cuda type and back? I would be happy to submit a PR if you think it'd be worthwhile?

**Comment by soumith** [comment]
i think that's worth adding, and its great that it wont crush the backward-compatibility. The naming seems okay too.
